## Segmenting and Clustering Neighborhoods in Toronto

For my assignment, I have used Beautiful Soup to scrape the Wikipedia page, and a combination of numpy and pandas to process the data into the required format. 

Happy to hear any useful tips and tricks!


In [1]:
#Install packages for beautifulsoup. Since i run Jupyter from anaconda these are already installed

! pip install beautifulsoup4 
! pip install request 

In [2]:
#import BeautifulSoup library and MAKE SURE TO IMPORT REQUESTS
from bs4 import BeautifulSoup as bs
import requests

In [3]:
#define a variable "source" which is the text of the html code for the wikipedia page. 
#Define "soup" as the text formatted to lxml.
#use the "prettify" function to reproduce the indents in the html file (and make it easier to read :)

source=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup=bs(source,'lxml')

I inspect the elements on the actual web page to work out which HTML element contains the body of the table

I discover it is the "tbody" element

So I specify a new variable "tab" which will find the first (and only!) instance of this element.

In [4]:
tab = soup.find('tbody').text
print(tab)


Postcode
Borough
Neighbourhood


M1A
Not assigned
Not assigned


M2A
Not assigned
Not assigned


M3A
North York
Parkwoods


M4A
North York
Victoria Village


M5A
Downtown Toronto
Harbourfront


M5A
Downtown Toronto
Regent Park


M6A
North York
Lawrence Heights


M6A
North York
Lawrence Manor


M7A
Queen's Park
Not assigned


M8A
Not assigned
Not assigned


M9A
Etobicoke
Islington Avenue


M1B
Scarborough
Rouge


M1B
Scarborough
Malvern


M2B
Not assigned
Not assigned


M3B
North York
Don Mills North


M4B
East York
Woodbine Gardens


M4B
East York
Parkview Hill


M5B
Downtown Toronto
Ryerson


M5B
Downtown Toronto
Garden District


M6B
North York
Glencairn


M7B
Not assigned
Not assigned


M8B
Not assigned
Not assigned


M9B
Etobicoke
Cloverdale


M9B
Etobicoke
Islington


M9B
Etobicoke
Martin Grove


M9B
Etobicoke
Princess Gardens


M9B
Etobicoke
West Deane Park


M1C
Scarborough
Highland Creek


M1C
Scarborough
Rouge Hill


M1C
Scarborough
Port Union


M2C
Not assigned
Not assigned


I discover this is a string separated by tabs. Oh my

I import pandas and numpy, and a new one called "re" and get straight to work

In [5]:
import pandas as pd
import numpy as np

In [6]:
tablist = tab.split('\t') #I am splitting the string into an list based on tab (symbolized by "'\t'")
tablist

["\nPostcode\nBorough\nNeighbourhood\n\n\nM1A\nNot assigned\nNot assigned\n\n\nM2A\nNot assigned\nNot assigned\n\n\nM3A\nNorth York\nParkwoods\n\n\nM4A\nNorth York\nVictoria Village\n\n\nM5A\nDowntown Toronto\nHarbourfront\n\n\nM5A\nDowntown Toronto\nRegent Park\n\n\nM6A\nNorth York\nLawrence Heights\n\n\nM6A\nNorth York\nLawrence Manor\n\n\nM7A\nQueen's Park\nNot assigned\n\n\nM8A\nNot assigned\nNot assigned\n\n\nM9A\nEtobicoke\nIslington Avenue\n\n\nM1B\nScarborough\nRouge\n\n\nM1B\nScarborough\nMalvern\n\n\nM2B\nNot assigned\nNot assigned\n\n\nM3B\nNorth York\nDon Mills North\n\n\nM4B\nEast York\nWoodbine Gardens\n\n\nM4B\nEast York\nParkview Hill\n\n\nM5B\nDowntown Toronto\nRyerson\n\n\nM5B\nDowntown Toronto\nGarden District\n\n\nM6B\nNorth York\nGlencairn\n\n\nM7B\nNot assigned\nNot assigned\n\n\nM8B\nNot assigned\nNot assigned\n\n\nM9B\nEtobicoke\nCloverdale\n\n\nM9B\nEtobicoke\nIslington\n\n\nM9B\nEtobicoke\nMartin Grove\n\n\nM9B\nEtobicoke\nPrincess Gardens\n\n\nM9B\nEtobicoke\

My list has only one (very long) element. But i notice the delimiter "\n" is being used. So i do one more split to get a multi-element list

In [7]:
tablist2=tablist[0].split('\n') 
len(tablist2)
print(tablist2)

['', 'Postcode', 'Borough', 'Neighbourhood', '', '', 'M1A', 'Not assigned', 'Not assigned', '', '', 'M2A', 'Not assigned', 'Not assigned', '', '', 'M3A', 'North York', 'Parkwoods', '', '', 'M4A', 'North York', 'Victoria Village', '', '', 'M5A', 'Downtown Toronto', 'Harbourfront', '', '', 'M5A', 'Downtown Toronto', 'Regent Park', '', '', 'M6A', 'North York', 'Lawrence Heights', '', '', 'M6A', 'North York', 'Lawrence Manor', '', '', 'M7A', "Queen's Park", 'Not assigned', '', '', 'M8A', 'Not assigned', 'Not assigned', '', '', 'M9A', 'Etobicoke', 'Islington Avenue', '', '', 'M1B', 'Scarborough', 'Rouge', '', '', 'M1B', 'Scarborough', 'Malvern', '', '', 'M2B', 'Not assigned', 'Not assigned', '', '', 'M3B', 'North York', 'Don Mills North', '', '', 'M4B', 'East York', 'Woodbine Gardens', '', '', 'M4B', 'East York', 'Parkview Hill', '', '', 'M5B', 'Downtown Toronto', 'Ryerson', '', '', 'M5B', 'Downtown Toronto', 'Garden District', '', '', 'M6B', 'North York', 'Glencairn', '', '', 'M7B', 'Not a

Now we are cooking! Let me first check the length of my new list

In [8]:
len(tablist2)

1446

I notice that there is a space before "Postcode" and one after "neighbourhood", so in fact instead of 3 columns, i need to re-shape my data to 5 columns.

But 1446 is not divisible by 5?

So think, why not just see what happens if i define an array based on 1445 elements

In [9]:
x=np.array(tablist2[0:1445])

I do a bit of math and work out if i have 5 columns, i must have 289 rows (to get to 1445 elements). So i define a new array based on re-shaping x

In [10]:
y=x.reshape(289,5)
y.shape

(289, 5)

In [11]:
y

array([['', 'Postcode', 'Borough', 'Neighbourhood', ''],
       ['', 'M1A', 'Not assigned', 'Not assigned', ''],
       ['', 'M2A', 'Not assigned', 'Not assigned', ''],
       ...,
       ['', 'M8Z', 'Etobicoke', 'Royal York South West', ''],
       ['', 'M8Z', 'Etobicoke', 'South of Bloor', ''],
       ['', 'M9Z', 'Not assigned', 'Not assigned', '']], dtype='<U49')

Now I define a dataframe based on this array

In [12]:
df=pd.DataFrame(data=y[1:,0:], columns=y[0,0:])

Next step: Work out which rows have Borough being not assigned

In [13]:
indexNames = df[ df['Borough'] =='Not assigned'].index

In [14]:
indexNames

Int64Index([  0,   1,   9,  13,  20,  21,  30,  36,  37,  45,  46,  50,  51,
             52,  54,  55,  59,  60,  61,  73,  74,  75,  88,  89,  90, 104,
            105, 106, 120, 121, 136, 137, 148, 149, 155, 161, 162, 167, 175,
            181, 182, 188, 189, 190, 194, 195, 201, 202, 203, 204, 209, 210,
            223, 224, 237, 238, 241, 242, 247, 248, 253, 254, 258, 259, 260,
            261, 263, 264, 274, 275, 276, 277, 278, 279, 280, 281, 287],
           dtype='int64')

Drop those rows

In [15]:
df.drop(indexNames , inplace=True)

For some reason I had to re-set the index because the new table did not automatically do that. so for example my first row had an index of 2?? Very weird

In [16]:
df=df.reset_index(drop=True)
df

Postcode           Borough             Neighbourhood   
0           M3A        North York                 Parkwoods   
1           M4A        North York          Victoria Village   
2           M5A  Downtown Toronto              Harbourfront   
3           M5A  Downtown Toronto               Regent Park   
4           M6A        North York          Lawrence Heights   
..  ..      ...               ...                       ... ..
206         M8Z         Etobicoke  Kingsway Park South West   
207         M8Z         Etobicoke                 Mimico NW   
208         M8Z         Etobicoke        The Queensway West   
209         M8Z         Etobicoke     Royal York South West   
210         M8Z         Etobicoke            South of Bloor   

[211 rows x 5 columns]

This next part was insane, and it took some Googling. 

Essentially there is an apply function that lets you group the column values. Don't ask me how it works

I do 3 things:

<ul>
  <li>Define a new dataframe which is <b> just </b> the postcodes with the neighbourhoods joined</li>
  <li>Define another dataframe which is the unique mapping of postcodes to boroughs</li>
  <li>Join these based on postcode</li>
</ul>


In [17]:
#Step 1

df2=df.groupby('Postcode')['Neighbourhood'].apply(lambda x: "%s" % ', '.join(x))
df2

Postcode
M1B                                       Rouge, Malvern
M1C               Highland Creek, Rouge Hill, Port Union
M1E                    Guildwood, Morningside, West Hill
M1G                                               Woburn
M1H                                            Cedarbrae
                             ...                        
M9N                                               Weston
M9P                                            Westmount
M9R    Kingsview Village, Martin Grove Gardens, Richv...
M9V    Albion Gardens, Beaumond Heights, Humbergate, ...
M9W                                            Northwest
Name: Neighbourhood, Length: 103, dtype: object

In [18]:
#Step 2

df3=df[['Postcode','Borough']].drop_duplicates()

In [19]:
#Step 3
result = pd.merge(df3,df2, on='Postcode')
result

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
..       ...               ...   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
3                     Lawrence Heights, Lawrence Manor  
4                                         Not assigned  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business Reply Mail Processing Centre 969 Eastern  
101  Humber Bay, King's Mill Park, Kingsway Park So...  
102  Kingsway Park South West, Mimico NW, The Queen...  

[103 rows x 3 columns]

I am going to be a stickler here and rename the postcode to Postal Code

In [20]:
result.rename(columns={"Postcode":"Postal Code"}, inplace=True)

Notice now there is a neighbourhood which is "not assigned".

We need to replace this with the "borough" name

How do we do this?

Approximately 2 hours of Google later, here is what i found.

There is a numpy function called <b> numpy.where </b>
    
Watch this

In [21]:
result['Neighbourhood'] = np.where(result['Neighbourhood'] == "Not assigned", result['Borough'], result['Neighbourhood'])

In [22]:
result

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A      Queen's Park   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
3                     Lawrence Heights, Lawrence Manor  
4                                         Queen's Park  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business Reply Mail Processing Centre 969 Eastern  
101  Humber Bay, King's Mill Park, Kingsway Park So...  
102  Kingsway Park South West, Mimico NW, The Queen...  

[103 rows x 3 columns]

<b> BOOM </b> 

Job done

Now lets add the co-ordinates

In [23]:
coord=pd.read_csv('https://cocl.us/Geospatial_data')  

In [24]:
coord.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [25]:
result_f = pd.merge(result,coord, on='Postal Code', how="left")
result_f.head()

Postal Code           Borough                     Neighbourhood   Latitude  \
0         M3A        North York                         Parkwoods  43.753259   
1         M4A        North York                  Victoria Village  43.725882   
2         M5A  Downtown Toronto         Harbourfront, Regent Park  43.654260   
3         M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4         M7A      Queen's Park                      Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

Define a new dataframe to be used as the basis for clustering the results by lat and longitude

In [26]:
result_clustering = result_f.drop(['Postal Code','Borough','Neighbourhood'], 1)
result_clustering

Latitude  Longitude
0    43.753259 -79.329656
1    43.725882 -79.315572
2    43.654260 -79.360636
3    43.718518 -79.464763
4    43.662301 -79.389494
..         ...        ...
98   43.653654 -79.506944
99   43.665860 -79.383160
100  43.662744 -79.321558
101  43.636258 -79.498509
102  43.628841 -79.520999

[103 rows x 2 columns]

Run k means to cluster by location. Display the results

In [27]:
#import kMeans
from sklearn.cluster import KMeans

kclusters=5

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(result_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 2, 3, 2, 1, 0, 4, 4, 2])

Insert Cluster latels

In [28]:
result_f.insert(0, 'Cluster Labels', kmeans.labels_)
result_f

Cluster Labels Postal Code           Borough  \
0                 4         M3A        North York   
1                 4         M4A        North York   
2                 2         M5A  Downtown Toronto   
3                 3         M6A        North York   
4                 2         M7A      Queen's Park   
..              ...         ...               ...   
98                1         M8X         Etobicoke   
99                2         M4Y  Downtown Toronto   
100               4         M7Y      East Toronto   
101               1         M8Y         Etobicoke   
102               1         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Harbourfront, Regent Park  43.654260 -79.360636  
3                     Lawrence Heights, Lawrence Manor  43.718518 -79.464763  
4                                         Queen's Park  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558  
101  Humber Bay, King's Mill Park, Kingsway Park So...  43.636258 -79.498509  
102  Kingsway Park South West, Mimico NW, The Queen...  43.628841 -79.520999  

[103 rows x 6 columns]

Import folium and matplotlib to generate map

In [29]:
import folium # map rendering library

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [30]:
# create map
map_clusters = folium.Map(location=[43.7, -79.3832], zoom_start=11)

In [31]:
# set color scheme for the clusters
xs = np.arange(kclusters)
ys = [i + xs + (i*xs)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

In [32]:
#define and generate map

for lat, lon, poi, cluster in zip(result_f['Latitude'], result_f['Longitude'], result_f['Neighbourhood'], result_f['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)

map_clusters